guardian/ophan-housekeeper
# GuCDK Housekeeper

## What does this change?

This PR migrates Ophan Housekeeper to GuCDK which is tracked in [#5086](https://github.com/guardian/ophan/issues/5086). The previous migrations, [Time Machine ](https://github.com/guardian/ophan/pull/5126), [Dashboard](https://github.com/guardian/ophan/pull/5146) and [Poller](https://github.com/guardian/ophan/pull/5174), we used `GuEc2App` but Housekeeper is a lambda and that is why we used `GuLambdaFunction`. 

We paired with @akash1810 in the DevX team as this was our first use of Lambda in GuCDK.

## How to test

We will update AWS CloudFormation with the GuCDK generated CFN and merge to main then we are going to compare the [logs in ELK](https://logs.gutools.co.uk/s/ophan/app/discover#/?_a=(columns:!(message,user.id),filters:!(),index:'581f7b00-c4df-11e9-bc08-9d1af4d9c1d5',interval:auto,query:(language:kuery,query:'%22running%20with%20notification%20message%22'),sort:!(!('@timestamp',desc)))&_g=(filters:!(),refreshInterval:(pause:!f,value:10000),time:(from:now-12h,to:now))) with the [CloudWatch metrics for the SES](https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#metricsV2:graph=~(metrics~(~(~'AWS*2fSES~'Bounce)~(~'.~'Delivery~(visible~false))~(~'.~'Reputation.BounceRate~(visible~false))~(~'.~'Reputation.ComplaintRate~(stat~'Average~visible~false))~(~'.~'Send~(stat~'Average~visible~false)))~view~'timeSeries~stacked~false~region~'eu-west-1~stat~'Sum~period~300~start~'-PT12H~end~'P0D);query=~'*7bAWS*2fSES*7d*20SES) and they are suppose to match if things are working well otherwise we will see bounces in CloudWatch that doesn't appear in our logs. We should be able to know the following day if we have a problem or things work as expected.

## How can we measure success?

We have the Ophan Housekeeper using GuCDK!

## Have we considered potential risks?

The risk is very low in here because the last time The Guardian was in an SES `probation` period, which means being added to a deny list and preventing The Guardian from sending emails, happened when the Lambda wasn't working for almost a month (for more info check this [PR](https://github.com/guardian/ophan/issues/2765)) but in this PR case we will check the following day if everything works as it should be and if not we will figure out what is missing and redo the same process again. One last thing to mention is AWS will notify us before we end up in `probation` and we have a grace period to resolve the issue - check [AWS SES docs](https://aws.amazon.com/blogs/messaging-and-targeting/what-if-you-get-an-ses-probation-notice-for-bounces/)
